### PR TITLE
enrich memberOf with options to just listen on items or groups

### PR DESCRIPTION
**Pull request by querdenker2k**
@JRuleWhenItemChange(item = GROUP_ITEM, memberOf = true)
changed to
@JRuleWhenItemChange(item = GROUP_ITEM, memberOf = JRuleMemberOf.All)
or
@JRuleWhenItemChange(item = GROUP_ITEM, memberOf = JRuleMemberOf.Items)
or
@JRuleWhenItemChange(item = GROUP_ITEM, memberOf = JRuleMemberOf.Groups)

In some of my rules I have something like that:
```
if (JRuleItem.forName(event.getMemberName()).isGroup()) {
    return;
}
```
To not do this inside the rule method, the rule should just be triggered for Items (and not Groups).

Further on the triggered itemName will be in itemName and not in memberName anymore, because https://github.com/seaside1/jrule/issues/93

### CODE REVIEW

**Comment by seime** [comment]
I like the idea. 
And would like to propose a similar extension to listen for events for members in subgroups. Ie 
`ITEM_A` memberOf `GROUP_A` memberOf `GROUP_B`

It would be great if I could get changes about `ITEM_A` if I listen for changes in `GROUP_B` even if `ITEM_A` isn't a direct member of `GROUP_B` but only indirectly via `GROUP_A`

Use case: Motion sensing in different zones of the premises. If I have multiple sensors in `GROUP_A` and the group status _isn't_ changed, I do not _think_ listening for memberOf events in `GROUP_B` triggers.

**Comment by seime** [comment]
Also remember to update the README with updated/new examples :)

**Comment by querdenker2k** [comment]
> I like the idea. And would like to propose a similar extension to listen for events for members in subgroups. Ie `ITEM_A` memberOf `GROUP_A` memberOf `GROUP_B`
> 
> It would be great if I could get changes about `ITEM_A` if I listen for changes in `GROUP_B` even if `ITEM_A` isn't a direct member of `GROUP_B` but only indirectly via `GROUP_A`
> 
> Use case: Motion sensing in different zones of the premises. If I have multiple sensors in `GROUP_A` and the group status _isn't_ changed, I do not _think_ listening for memberOf events in `GROUP_B` triggers.

Thought about the same and started doing it but reverted it again. Maybe we could this in an additional MR. I have to much open MRs and I have to merge them all again in my own developer branch. This should be simply from the user view with additional enums for JRuleMemberOf

**Comment by seime** [comment]
@querdenker2k could you rebase and resolve conflicts?

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java [hunk 1]
```diff
@@ -49,74 +50,110 @@ public void initTestClass() throws ItemNotFoundException {
...
-            return stringItem;
-        });
-
-        JRuleGroupItemChangeRules rule = initRule(JRuleGroupItemChangeRules.class);
-        // Only last event should trigger rule method
-        fireEvents(List.of(itemChangeEvent(MEMBER_ITEM, "2", "1"), itemChangeEvent(MEMBER_ITEM, "3", "2"),
-                itemChangeEvent(MEMBER_ITEM, "1", "2")));
-        verify(rule, times(1)).groupItemChangeFromTo(Mockito.any(JRuleEvent.class));
-    }
+//    @Test
```
Comment (seime) [inline src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java:94]: A lot of commented out tests - should these be enabled again?

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java [hunk 2]
```diff
@@ -49,74 +50,110 @@ public void initTestClass() throws ItemNotFoundException {
...
-            Mockito.when(stringItem.getGroupNames()).thenReturn(List.of(JRuleGroupItemChangeRules.GROUP_ITEM_FROM));
-            return stringItem;
-        });
-
-        JRuleGroupItemChangeRules rule = initRule(JRuleGroupItemChangeRules.class);
-        // Only last event should trigger rule method
-        fireEvents(List.of(itemChangeEvent(MEMBER_ITEM, "2", "1"), itemChangeEvent(MEMBER_ITEM, "1", "2")));
-        verify(rule, times(1)).groupItemChangeFrom(Mockito.any(JRuleEvent.class));
-    }
+//    @Test
```
Comment (seime) [inline src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java:53]: Same here - obsolete or temporarily commented out while waiting for another PR to merge?

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeRules.java [hunk 3]
```diff
@@ -26,27 +27,39 @@
...
+    @JRuleName("Test JRuleWhenGroupItemChange/JustItems")
+    @JRuleWhenItemChange(item = GROUP_ITEM_JUST_ITEMS, memberOf = JRuleMemberOf.Items)
+    public void groupItemChangeJustItems(JRuleEvent event) {
     }
 
-    @JRuleName("Test JRuleWhenGroupItemChange/from/to")
-    @JRuleWhenItemChange(item = GROUP_ITEM_FROM_TO, from = "1", to = "2", memberOf = true)
-    public void groupItemChangeFromTo(JRuleEvent event) {
-    }
+//    @JRuleName("Test JRuleWhenGroupItemChange/JustGroups")
```
Comment (seime) [inline src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeRules.java:46]: Same

**Review by seime — CHANGES_REQUESTED** [review]
Is this PR still a draft or ready for review?

**Comment by querdenker2k** [comment]
not ready yet, have to resolve merge conflicts (therefor the outcommented tests for better debugging) but will not get it to work in the next 5 days.

**Comment by querdenker2k** [comment]
Ready again, added IT tests.
Currently this is working just for direct members of groups (as before).
Additionaly an *_Recursive could be added to the enum.

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeRules.java (hunk 3 above)
Comment (querdenker2k) [reply src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeRules.java]: uncommented

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java (hunk 2 above)
Comment (querdenker2k) [reply src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java]: uncommented

File: src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java (hunk 1 above)
Comment (querdenker2k) [reply src/test/java/org/openhab/binding/jrule/internal/triggers/itemchange/JRuleGroupItemChangeTest.java]: uncommented

**Comment by querdenker2k** [comment]
PUSH

File: README.md [hunk 4]
```diff
@@ -718,11 +718,14 @@ public void testPower(JRuleEvent event) {
...
 ## Example 35
 
-Use case: Want to listen on all Item events of a group (without the groupstate must change)
+Use case: Want to listen on all Item events of a group (without the groupstate must change). 
+    Alternatively you could just listen to just Group changes or (real) Item changes
 
 ```java
     @JRuleName("MemberOfUpdateTrigger")
-    @JRuleWhenItemReceivedUpdate(item = _MySwitchGroup.ITEM, memberOf = true)
+    @JRuleWhenItemReceivedUpdate(item = _MySwitchGroup.ITEM, memberOf = JRuleMemberOf.All)
```
Comment (seime) [inline README.md:726]: Minor - commented out code

**Review by seime — APPROVED** [review]
Not tested, but has unit tests

**Comment by seime** [comment]
@querdenker2k  Tested this a bit tonight, and something is not correct.

I replaced all my `memberOf = true` with `memberOf = JRuleMemberOf.All` as indicated in the README file.

When using `JRuleMemberOf.All`, rules listening for changes in a specific group suddenly gets triggered by changes happening to items in _unrelated_ groups.

Here is one example;
CleanShot 2023-02-11 at 19 15 48

CleanShot 2023-02-11 at 19 16 11

The event claims to be for group `gFloodSensors`, but the member is part of `gMovementSensorsHouse`. In fact the `gFloodSensors` group does not have any members at the moment, and the groups are unrelated. 

I have also seen my batteryGroup monitoring rules suddenly getting triggered with motion sensor events etc.

Not sure where the problem is; either parsing of which groups contains which members or the rule filtering mechanism.

Changing from `All` to `Items` seem to stop this erratic behaviour, but not really sure if this makes it correct.

**Comment by querdenker2k** [comment]
Damn, will look into it. Could already create a negative test which describes your behaviour.

**Comment by querdenker2k** [comment]
fixed this, problem was very stupid

**Comment by seime** [comment]
Appears to be working as expected now!

Thanks @querdenker2k

**Review by seime — APPROVED** [review]

**Pull request merged**